### ISSUE TITLE
raise a warning if someone tries to access .grad of a Variable that will never be computed

### ISSUE THREAD

**Issue by BestSonny** [pytorch/pytorch#2362]
@soumith 
In my understanding, these two kinds of initializations should give the same result.

```
grid_data_variable = Variable(torch.rand(batch_size, gh, gw, d, nchans), requires_grad=True).cuda()
guide_data_variable = Variable(torch.rand(batch_size, h, w), requires_grad=True).cuda()
```

```
grid_data_variable = Variable(torch.rand(batch_size, gh, gw, d, nchans).cuda(), requires_grad=True)
guide_data_variable = Variable(torch.rand(batch_size, h, w).cuda(), requires_grad=True)
```

Then I test them on my customized layer.
The first one would fail in the **gradcheck** while the second works well.

Do you have any clues why the first one gives an incorrect gradient?

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by albanD** [comment]
You error is that in the first case, what you store in `grid_data_variable` and `guide_data_variable` is not the Variable with `requires_grad=True` but the Variable moved to `.cuda()` for which gradients won't be computed and thus the gradcheck won't be correct.

**Comment by BestSonny** [comment]
@albanD A bit confused. I checked the first case, both the ```grid_data_variable``` and ```guide_data_variable``` are ```True``` on the ```requires_grad```.

Still, I can not understand why the gradient won't be computed.

**Comment by albanD** [comment]
The gradients are computed only for Variable that are **created by the user** and with **`requires_grad=True`**.
In your case, the Variable that you store is not the one that is created:
 ```
# Your first line can be separated into these two lines
user_grid_data_variable = Variable(torch.rand(batch_size, gh, gw, d, nchans), requires_grad=True)
grid_data_variable = user_grid_data_variable.cuda()
# user_grid_data_variable will have gradients properly computed
# grid_data_variable  will not have gradients as it is not user created, it has been created by the `.cuda()` operation

```

**Comment by gchanan** [comment]
I've noticed a number of people run into this issue; I wonder if there's something we can do to make it more obvious what is going on -- I can't think of anything off the top of my head.

**Comment by apaszke** [comment]
We could raise a warning if someone tries to access .grad of a Variable that will never be computed
